applications: Add Asset Tracker v2

jtguggedal opened the pull request:
Add Asset Tracker v2 to NCS.

jhn-nordic commented on applications/asset_tracker_v2/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright (c) 2019-2020 Nordic Semiconductor ASA
```
2021

jhn-nordic replied:
In a lot of files

jtguggedal replied:
Fixed

jtguggedal commented on applications/asset_tracker_v2/Kconfig:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+menu "Cat Tracker"
```
Fixed

jtguggedal commented on applications/asset_tracker_v2/src/modules/Kconfig.cloud_module:
```diff
@@ -0,0 +1,46 @@
+config MQTT_CLIENT_ID
+	depends on USE_CUSTOM_MQTT_CLIENT_ID
+	string	"Configures the custom MQTT client ID"
+	default "yoghurt"
```
Removed default, and set in `prj_*.conf` instead

jtguggedal commented on applications/asset_tracker_v2/nrf9160dk_nrf9160ns.overlay:
```diff
@@ -0,0 +1,15 @@
+&spi3 {
```
Removed overlay and reference to it in the docs. It`s not needed for DK

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-3746
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

coderbyheart commented:
CLangFormat reports some issues, are we still honoring that style?
https://github.com/NordicSemiconductor/asset-tracker-cloud-firmware/runs/1830991902?check_suite_focus=true

jtguggedal commented:
> CLangFormat reports some issues, are we still honoring that style?
> https://github.com/NordicSemiconductor/asset-tracker-cloud-firmware/runs/1830991902?check_suite_focus=true

I can't quite see why ClangFormat fails, and I have only checked that checkpatch is happy with the formatting. But I think we should be able to comply with the ClangFormat rules as well if they don't conflict with the NCS checkpatch rules.

coderbyheart commented:
The test is using the NCS rules: https://github.com/NordicSemiconductor/asset-tracker-cloud-firmware/blob/ncs-v1.5/.github/workflows/pr-clang-format.yaml#L25

jtguggedal commented:
It looks like the ClangFormat test will fail as long as ClangFormat reports a diff. If that's the case, then I think we will not aim to pass the Clang test as long as the CI's checkpatch passes, as I find some of the suggested code in the diff less readable and not in line with convention in NCS. Maybe it's possible to tweak the Clang parameters to expect code that is more in line with the convention.

coderbyheart commented:
Yeah, the advantage formatting code to a specific style is that it avoids getting white-space diffs because different developers use different styles. But since it seems that in NCS code is generally *not* automatically formatted, we can ignore this check (and I will make it not fail, but keep it in the repo for customers).

mbolivar-nordic reviewed (commented):
I gave this a quick look and it looks very nice. I noticed a couple of things here and there.

carlescufi commented on applications/asset_tracker_v2/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Indeed it does, I just sent an email about this.

jtguggedal replied:
Thanks, I've updated the license in all new files introduced in the PR

jtguggedal commented:
@mbolivar-nordic Thanks for having a look! I've addressed your comments

jtguggedal commented:
The CI failure is being fixed in https://github.com/nrfconnect/sdk-nrf/pull/3820, so merging that PR is a requirement to make CI green in this PR.

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_codec/aws_iot_codec.c:
```diff
@@ -0,0 +1,821 @@
+		goto exit;
+	}
+
+	buffer = cJSON_PrintUnformatted(root_obj);
```
Shouldn't it be checked if `buffer == NULL`? `cJSON_PrintUnformatted()` seems to use dynamic memory allocation, if it fails we'll end up with a NULL pointer, which may crash the app (in the `strlen()` call below for instance).

Same would apply in other places in this file where `cJSON_PrintUnformatted()` is called.

jtguggedal replied:
Indeed, fixed buffer check and cleanup in case of failure for all occurences of `cJSON_PrintUnformatted()`

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_codec/cloud_codec.h:
```diff
@@ -0,0 +1,232 @@
+
+static inline void cloud_codec_release_data(struct cloud_codec_data *output)
+{
+	k_free(output->buf);
```
There's a dedicated `cJSON_FreeString()` function to release buffers allocated by CJSON lib.

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_codec/cloud_codec_ringbuffer.c:
```diff
@@ -0,0 +1,138 @@
+				   int *head_ui_buf,
+				   size_t buffer_count)
+{
+
```
nit: Unneeded empty line.

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_codec/cloud_codec_ringbuffer.c:
```diff
@@ -0,0 +1,138 @@
+#include <cloud_codec.h>
```
Missing license info.

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_codec/json_aux.c:
```diff
@@ -0,0 +1,76 @@
+#include <errno.h>
```
Missing license info.

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_codec/json_aux.h:
```diff
@@ -0,0 +1,15 @@
+#include "cJSON.h"
```
Missing license info.

rlubos commented on applications/asset_tracker_v2/src/cloud/cloud_wrapper.h:
```diff
@@ -0,0 +1,67 @@
+#include <zephyr.h>
```
Missing license info.

rlubos commented on applications/asset_tracker_v2/src/ext_sensors/ext_sensors.c:
```diff
@@ -0,0 +1,263 @@
+	.dev_name = CONFIG_ACCELEROMETER_DEV_NAME
+};
+
+static ext_sensor_handler_t m_evt_handler;
```
nit: Hungarian notation shouldn't be used.

rlubos commented on applications/asset_tracker_v2/src/ext_sensors/ext_sensors.c:
```diff
@@ -0,0 +1,263 @@
+#include <zephyr.h>
```
Missing license info.

rlubos commented on applications/asset_tracker_v2/src/modules/sensor_module.c:
```diff
@@ -0,0 +1,335 @@
+static void module_thread_fn(void)
+{
+	int err;
+
```
nit: Unneeded empty line.

rlubos reviewed (commented):
Quite massive, but neat. A few minor issues spotted.

jtguggedal commented:
@rlubos Thanks for taking time to review this giant PR! I think I have addressed all of your comments, along with some cleanup of repeated code. Diff here: https://github.com/nrfconnect/sdk-nrf/compare/5ef68c278f042094a75cce94346695229c95f3a1..caeeefe5b90b223b49fb06945e13de1defdfd786

(cc @simensrostad)

rlubos reviewed (approved):
Thanks

coderbyheart commented:
`CONFIG_LOG_MAX_LEVEL=4` seems not to enable all debug logs, this is how the log output looks like:

```
*** Booting Zephyr OS version 2.4.99  ***e slot
/dev/ttyACM0 [00:00:00.212,463] <inf> event_manager: APP_EVT_START
/dev/ttyACM0 [00:00:00.226,562] <inf> event_manager: DATA_EVT_CONFIG_INIT
/dev/ttyACM0 [00:00:00.233,001] <inf> app_module: Device mode: Active
/dev/ttyACM0 [00:00:00.238,769] <inf> app_module: Start data sample timer: 120 seconds interval
/dev/ttyACM0 [00:00:00.298,034] <inf> event_manager: MODEM_EVT_LTE_CONNECTING
/dev/ttyACM0 [00:00:26.568,481] <inf> event_manager: MODEM_EVT_LTE_DISCONNECTED
/dev/ttyACM0 [00:00:26.575,378] <inf> event_manager: MODEM_EVT_LTE_CELL_UPDATE
/dev/ttyACM0 [00:00:28.987,518] <inf> event_manager: MODEM_EVT_LTE_CONNECTED
/dev/ttyACM0 [00:00:28.994,171] <inf> event_manager: MODEM_EVT_LTE_PSM_UPDATE
/dev/ttyACM0 [00:00:29.000,976] <wrn> cloud_module: Cloud connection establishment in progress
/dev/ttyACM0 [00:00:29.009,185] <wrn> cloud_module: New connection attempt in 32 seconds if not successful
/dev/ttyACM0 [00:00:29.018,676] <inf> event_manager: CLOUD_EVT_CONNECTING
/dev/ttyACM0 [00:00:33.543,518] <inf> event_manager: CLOUD_EVT_DISCONNECTED
/dev/ttyACM0 [00:00:45.448,699] <inf> event_manager: MODEM_EVT_LTE_CELL_UPDATE
/dev/ttyACM0 [00:01:01.018,707] <inf> event_manager: CLOUD_EVT_CONNECTION_TIMEOUT
/dev/ttyACM0 [00:01:01.025,756] <wrn> cloud_module: Cloud connection establishment in progress
/dev/ttyACM0 [00:01:01.033,966] <wrn> cloud_module: New connection attempt in 64 seconds if not successful
/dev/ttyACM0 [00:01:01.043,457] <inf> event_manager: CLOUD_EVT_CONNECTING
/dev/ttyACM0 [00:01:03.672,851] <inf> event_manager: CLOUD_EVT_CONNECTED
/dev/ttyACM0 [00:01:03.934,082] <inf> event_manager: DATA_EVT_DATE_TIME_OBTAINED
/dev/ttyACM0 [00:01:03.941,131] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_STAT, MOD_DYN, BAT, ENV)
/dev/ttyACM0 [00:01:03.951,965] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
/dev/ttyACM0 [00:01:03.993,682] <wrn> modem_module: Unsupported modem firmware version: mfw_nrf9160_1.2.2
/dev/ttyACM0 [00:01:04.002,868] <wrn> modem_module: Expected firmware version: mfw_nrf9160_1.2.3
/dev/ttyACM0 [00:01:04.011,352] <wrn> modem_module: You can change the expected version through the
/dev/ttyACM0 [00:01:04.020,080] <wrn> modem_module: EXPECTED_MODEM_FIRMWARE_VERSION setting.
/dev/ttyACM0 [00:01:04.028,198] <wrn> modem_module: Please upgrade: http://bit.ly/nrf9160-mfw-update
/dev/ttyACM0 [00:01:04.037,078] <inf> event_manager: MODEM_EVT_MODEM_STATIC_DATA_READY
/dev/ttyACM0 [00:01:04.078,430] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
/dev/ttyACM0 [00:01:04.120,147] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
/dev/ttyACM0 [00:01:04.127,258] <inf> event_manager: DATA_EVT_DATA_READY
/dev/ttyACM0 [00:01:04.141,204] <inf> event_manager: DATA_EVT_DATA_SEND
/dev/ttyACM0 [00:01:04.147,644] <wrn> data_module: No batch data to encode, ringbuffers empty
/dev/ttyACM0 [00:01:04.158,264] <inf> event_manager: CLOUD_EVT_DATA_ACK
/dev/ttyACM0 [00:02:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
/dev/ttyACM0 [00:02:00.253,540] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
/dev/ttyACM0 [00:02:00.284,667] <inf> event_manager: GPS_EVT_ACTIVE
/dev/ttyACM0 [00:02:00.290,618] <inf> event_manager: GPS_EVT_AGPS_NEEDED
/dev/ttyACM0 [00:02:00.297,760] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
/dev/ttyACM0 [00:02:00.338,439] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
/dev/ttyACM0 [00:02:00.379,455] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
/dev/ttyACM0 [00:03:04.844,482] <inf> event_manager: GPS_EVT_TIMEOUT
/dev/ttyACM0 [00:03:04.850,280] <inf> event_manager: GPS_EVT_INACTIVE
/dev/ttyACM0 [00:03:04.856,353] <inf> event_manager: DATA_EVT_DATA_READY
/dev/ttyACM0 [00:03:04.866,577] <inf> event_manager: DATA_EVT_DATA_SEND
/dev/ttyACM0 [00:03:04.873,016] <wrn> data_module: No batch data to encode, ringbuffers empty
/dev/ttyACM0 [00:03:04.889,434] <inf> event_manager: CLOUD_EVT_DATA_ACK
/dev/ttyACM0 [00:04:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
/dev/ttyACM0 [00:04:00.253,509] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
/dev/ttyACM0 [00:04:00.270,568] <inf> event_manager: GPS_EVT_ACTIVE
/dev/ttyACM0 [00:04:00.276,550] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
/dev/ttyACM0 [00:04:00.317,871] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
/dev/ttyACM0 [00:04:00.358,703] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
/dev/ttyACM0 [00:05:04.804,962] <inf> event_manager: GPS_EVT_TIMEOUT
/dev/ttyACM0 [00:05:04.810,760] <inf> event_manager: GPS_EVT_INACTIVE
/dev/ttyACM0 [00:05:04.816,833] <inf> event_manager: DATA_EVT_DATA_READY
/dev/ttyACM0 [00:05:04.827,056] <inf> event_manager: DATA_EVT_DATA_SEND
/dev/ttyACM0 [00:05:04.833,496] <wrn> data_module: No batch data to encode, ringbuffers empty
/dev/ttyACM0 [00:05:04.849,945] <inf> event_manager: CLOUD_EVT_DATA_ACK
/dev/ttyACM0 [00:06:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
/dev/ttyACM0 [00:06:00.253,509] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
/dev/ttyACM0 [00:06:00.290,557] <inf> event_manager: GPS_EVT_ACTIVE
/dev/ttyACM0 [00:06:00.296,539] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
/dev/ttyACM0 [00:06:00.337,738] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
/dev/ttyACM0 [00:06:00.378,479] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
/dev/ttyACM0 [00:07:05.126,892] <inf> event_manager: GPS_EVT_TIMEOUT
/dev/ttyACM0 [00:07:05.132,690] <inf> event_manager: GPS_EVT_INACTIVE
/dev/ttyACM0 [00:07:05.138,763] <inf> event_manager: DATA_EVT_DATA_READY
/dev/ttyACM0 [00:07:05.149,017] <inf> event_manager: DATA_EVT_DATA_SEND
/dev/ttyACM0 [00:07:05.155,456] <wrn> data_module: No batch data to encode, ringbuffers empty
/dev/ttyACM0 [00:07:05.171,875] <inf> event_manager: CLOUD_EVT_DATA_ACK
/dev/ttyACM0 [00:08:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
/dev/ttyACM0 [00:08:00.253,509] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
/dev/ttyACM0 [00:08:00.270,507] <inf> event_manager: GPS_EVT_ACTIVE
/dev/ttyACM0 [00:08:00.276,702] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
/dev/ttyACM0 [00:08:00.318,176] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
/dev/ttyACM0 [00:08:00.359,283] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
```

coderbyheart commented:
I get an error when scheduling a FOTA:

```
/dev/ttyACM0 [00:01:07.152,679] <inf> event_manager: CLOUD_EVT_DATA_ACK
/dev/ttyACM0 [00:01:13.883,728] <err> aws_fota: Error (-7) when trying to start firmware download
```

Did anything of notice change?

mbolivar-nordic commented on applications/asset_tracker_v2/src/ext_sensors/ext_sensors.c:
```diff
@@ -0,0 +1,269 @@
+
+struct env_sensor {
+	enum sensor_channel channel;
+	uint8_t *dev_name;
```
Making this `const char*` would be clearer, no?

jtguggedal replied:
Agreed, fixed

mbolivar-nordic commented on applications/asset_tracker_v2/src/ext_sensors/ext_sensors.c:
```diff
@@ -0,0 +1,269 @@
+
+static struct env_sensor temp_sensor = {
+	.channel = SENSOR_CHAN_AMBIENT_TEMP,
+	.dev_name = CONFIG_MULTISENSOR_DEV_NAME
+};
+
+static struct env_sensor humid_sensor = {
+	.channel = SENSOR_CHAN_HUMIDITY,
+	.dev_name = CONFIG_MULTISENSOR_DEV_NAME
+};
+
+static struct env_sensor accel_sensor = {
+	.channel = SENSOR_CHAN_ACCEL_XYZ,
+	.dev_name = CONFIG_ACCELEROMETER_DEV_NAME
```
These dev_name values are all being initialized from string literals and therefore are in ROM, so there is no need to log_strdup() them in ext_sensors_mov_thres_set().

https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/reference/logging/index.html#logging-strings

As an (even better) alternative, you may be able to avoid saving device names at all by just using DEVICE_DT_GET() and dropping all of these DEV_NAME config options. This sort of thing is deprecated anyway: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/kconfig/tips.html#options-that-specify-a-device-in-the-system-by-name

jtguggedal replied:
I have removed the `log_strdup()`s from these and will look into fetching the device name from the device tree instead.

jtguggedal replied:
My device tree skills are quite limited, so I'm not entirely sure how to go about this. Is this in the right direction?
https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/f16fff1931955f28eb60310121caa04ed18b9a16

jtguggedal replied:
I'm using aliases so that other boards can more easily plug into the same code with a DT overlay

mbolivar-nordic replied:
> My device tree skills are quite limited, so I'm not entirely sure how to go about this. Is this in the right direction?
> [jtguggedal@f16fff1](https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/f16fff1931955f28eb60310121caa04ed18b9a16)
>
> I'm using aliases so that other boards can more easily plug into the same code with a DT overlay

Using aliases is fine, but you don't need to store the device label property at all anymore. You can just get the device straight away using DEVICE_DT_GET.

For example:

```c
struct env_sensor {
	enum sensor_channel channel;
	const struct device *dev;
	struct k_spinlock lock;
};

static struct env_sensor temp_sensor = {
	.channel = SENSOR_CHAN_AMBIENT_TEMP,
	.dev = DEVICE_DT_GET(DT_ALIAS(temp_sensor)),
};
```

Then you just have to check that `device_is_ready(temp_sensor.dev)` returns 0 in your init routine, and error out if it's not.

This has the following advantages:

- runs faster: no looping over devices in device_get_binding()
- fails faster: a missing device due to a bad devicetree configuration is a build error, not a runtime error
- smaller footprint: no dev_name

The name of the device is still accessible as `temp_sensor.dev->name` for debugging purposes.

jtguggedal replied:
Oh, so that's how it works, pretty neat! I updated per your suggestions above: https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/00002bf09031fbf6dbbd30bff061fa058067148

The alias stuff in `thingy91_nrf9160ns.overlay` felt a bit hacky to do this way, but it was the only way I found that worked. Is there a better way to alias a device on a given bus, like the BME680 or the ADXL362?

mbolivar-nordic replied:
> Thanks for explaining, these new DT macros offer some nice functionality that I wasn't aware of.

I strongly agree. Hat tip to @pabigot for making this happen for zephyr 2.5.

> Is there a better way to alias a device on a given bus, like the BME680 or the ADXL362

Depends on how you'd like to go about it. DEVICE_DT_GET() takes a devicetree node identifier, and there are a lot of ways to specify those.

I'm guessing you don't want to force the user to pick a BME680 for ambient temperature; e.g. if they have another sensor device that supplies SENSOR_CHAN_AMBIENT_TEMP, they should be able to use it easily by just tweaking their devicetree. And similarly for ADXL362.

Is that right? If so, you basically are limited either aliases or node labels.

Aliases you've already got a solution for.

If you want to use node labels, that'd look like this in the overlay:

```
&i2c2 {
	asset_tracker_temp: bme680@76 { ... };
};
```

Then you could say `DEVICE_DT_GET(DT_NODELABEL(asset_tracker_temp))` in C.

You could technically also use the `/zephyr,user` node (see the DT guide for documentation), but in this situation I think `/aliases` would be more idiomatic.

jtguggedal replied:
> I'm guessing you don't want to force the user to pick a BME680 for ambient temperature; e.g. if they have another sensor device that supplies SENSOR_CHAN_AMBIENT_TEMP, they should be able to use it easily by just tweaking their devicetree. And similarly for ADXL362.

Yes, that's what I want to do. Looks like what I have on https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/00002bf09031fbf6dbbd30bff061fa058067148 is fairly reasonable, then, so I'll submit a PR with that. Thanks for helping out on this.

mbolivar-nordic replied:
>  Looks like what I have on [jtguggedal@00002bf](https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/00002bf09031fbf6dbbd30bff061fa058067148) is fairly reasonable, then, so I'll submit a PR with that. Thanks for helping out on this.

Sounds good, and sure thing.

mbolivar-nordic commented on applications/asset_tracker_v2/src/main.c:
```diff
@@ -0,0 +1,481 @@
+static void state_set(enum state_type new_state)
+{
+	if (new_state == state) {
+		LOG_DBG("State: %s", log_strdup(state2str(state)));
```
I don't think any of these need to be given to log_strdup() either. The sub_state2str() function is returning string literals, which should be allocated in ROM.

jtguggedal replied:
You're right, they should be, so `log_strdup()` has been removed

mbolivar-nordic reviewed (commented):
Thanks for the update! I might be missing something (pinging @nordic-krch for confirmation), but I believe there are various additional log_strdup() calls that are not necessary and could be removed.

simensrostad commented:
> `CONFIG_LOG_MAX_LEVEL=4` seems not to enable all debug logs, this is how the log output looks like:
> 
> ```
> *** Booting Zephyr OS version 2.4.99  ***e slot
> /dev/ttyACM0 [00:00:00.212,463] <inf> event_manager: APP_EVT_START
> /dev/ttyACM0 [00:00:00.226,562] <inf> event_manager: DATA_EVT_CONFIG_INIT
> /dev/ttyACM0 [00:00:00.233,001] <inf> app_module: Device mode: Active
> /dev/ttyACM0 [00:00:00.238,769] <inf> app_module: Start data sample timer: 120 seconds interval
> /dev/ttyACM0 [00:00:00.298,034] <inf> event_manager: MODEM_EVT_LTE_CONNECTING
> /dev/ttyACM0 [00:00:26.568,481] <inf> event_manager: MODEM_EVT_LTE_DISCONNECTED
> /dev/ttyACM0 [00:00:26.575,378] <inf> event_manager: MODEM_EVT_LTE_CELL_UPDATE
> /dev/ttyACM0 [00:00:28.987,518] <inf> event_manager: MODEM_EVT_LTE_CONNECTED
> /dev/ttyACM0 [00:00:28.994,171] <inf> event_manager: MODEM_EVT_LTE_PSM_UPDATE
> /dev/ttyACM0 [00:00:29.000,976] <wrn> cloud_module: Cloud connection establishment in progress
> /dev/ttyACM0 [00:00:29.009,185] <wrn> cloud_module: New connection attempt in 32 seconds if not successful
> /dev/ttyACM0 [00:00:29.018,676] <inf> event_manager: CLOUD_EVT_CONNECTING
> /dev/ttyACM0 [00:00:33.543,518] <inf> event_manager: CLOUD_EVT_DISCONNECTED
> /dev/ttyACM0 [00:00:45.448,699] <inf> event_manager: MODEM_EVT_LTE_CELL_UPDATE
> /dev/ttyACM0 [00:01:01.018,707] <inf> event_manager: CLOUD_EVT_CONNECTION_TIMEOUT
> /dev/ttyACM0 [00:01:01.025,756] <wrn> cloud_module: Cloud connection establishment in progress
> /dev/ttyACM0 [00:01:01.033,966] <wrn> cloud_module: New connection attempt in 64 seconds if not successful
> /dev/ttyACM0 [00:01:01.043,457] <inf> event_manager: CLOUD_EVT_CONNECTING
> /dev/ttyACM0 [00:01:03.672,851] <inf> event_manager: CLOUD_EVT_CONNECTED
> /dev/ttyACM0 [00:01:03.934,082] <inf> event_manager: DATA_EVT_DATE_TIME_OBTAINED
> /dev/ttyACM0 [00:01:03.941,131] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_STAT, MOD_DYN, BAT, ENV)
> /dev/ttyACM0 [00:01:03.951,965] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
> /dev/ttyACM0 [00:01:03.993,682] <wrn> modem_module: Unsupported modem firmware version: mfw_nrf9160_1.2.2
> /dev/ttyACM0 [00:01:04.002,868] <wrn> modem_module: Expected firmware version: mfw_nrf9160_1.2.3
> /dev/ttyACM0 [00:01:04.011,352] <wrn> modem_module: You can change the expected version through the
> /dev/ttyACM0 [00:01:04.020,080] <wrn> modem_module: EXPECTED_MODEM_FIRMWARE_VERSION setting.
> /dev/ttyACM0 [00:01:04.028,198] <wrn> modem_module: Please upgrade: http://bit.ly/nrf9160-mfw-update
> /dev/ttyACM0 [00:01:04.037,078] <inf> event_manager: MODEM_EVT_MODEM_STATIC_DATA_READY
> /dev/ttyACM0 [00:01:04.078,430] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
> /dev/ttyACM0 [00:01:04.120,147] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
> /dev/ttyACM0 [00:01:04.127,258] <inf> event_manager: DATA_EVT_DATA_READY
> /dev/ttyACM0 [00:01:04.141,204] <inf> event_manager: DATA_EVT_DATA_SEND
> /dev/ttyACM0 [00:01:04.147,644] <wrn> data_module: No batch data to encode, ringbuffers empty
> /dev/ttyACM0 [00:01:04.158,264] <inf> event_manager: CLOUD_EVT_DATA_ACK
> /dev/ttyACM0 [00:02:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
> /dev/ttyACM0 [00:02:00.253,540] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
> /dev/ttyACM0 [00:02:00.284,667] <inf> event_manager: GPS_EVT_ACTIVE
> /dev/ttyACM0 [00:02:00.290,618] <inf> event_manager: GPS_EVT_AGPS_NEEDED
> /dev/ttyACM0 [00:02:00.297,760] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
> /dev/ttyACM0 [00:02:00.338,439] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
> /dev/ttyACM0 [00:02:00.379,455] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
> /dev/ttyACM0 [00:03:04.844,482] <inf> event_manager: GPS_EVT_TIMEOUT
> /dev/ttyACM0 [00:03:04.850,280] <inf> event_manager: GPS_EVT_INACTIVE
> /dev/ttyACM0 [00:03:04.856,353] <inf> event_manager: DATA_EVT_DATA_READY
> /dev/ttyACM0 [00:03:04.866,577] <inf> event_manager: DATA_EVT_DATA_SEND
> /dev/ttyACM0 [00:03:04.873,016] <wrn> data_module: No batch data to encode, ringbuffers empty
> /dev/ttyACM0 [00:03:04.889,434] <inf> event_manager: CLOUD_EVT_DATA_ACK
> /dev/ttyACM0 [00:04:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
> /dev/ttyACM0 [00:04:00.253,509] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
> /dev/ttyACM0 [00:04:00.270,568] <inf> event_manager: GPS_EVT_ACTIVE
> /dev/ttyACM0 [00:04:00.276,550] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
> /dev/ttyACM0 [00:04:00.317,871] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
> /dev/ttyACM0 [00:04:00.358,703] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
> /dev/ttyACM0 [00:05:04.804,962] <inf> event_manager: GPS_EVT_TIMEOUT
> /dev/ttyACM0 [00:05:04.810,760] <inf> event_manager: GPS_EVT_INACTIVE
> /dev/ttyACM0 [00:05:04.816,833] <inf> event_manager: DATA_EVT_DATA_READY
> /dev/ttyACM0 [00:05:04.827,056] <inf> event_manager: DATA_EVT_DATA_SEND
> /dev/ttyACM0 [00:05:04.833,496] <wrn> data_module: No batch data to encode, ringbuffers empty
> /dev/ttyACM0 [00:05:04.849,945] <inf> event_manager: CLOUD_EVT_DATA_ACK
> /dev/ttyACM0 [00:06:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
> /dev/ttyACM0 [00:06:00.253,509] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
> /dev/ttyACM0 [00:06:00.290,557] <inf> event_manager: GPS_EVT_ACTIVE
> /dev/ttyACM0 [00:06:00.296,539] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
> /dev/ttyACM0 [00:06:00.337,738] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
> /dev/ttyACM0 [00:06:00.378,479] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
> /dev/ttyACM0 [00:07:05.126,892] <inf> event_manager: GPS_EVT_TIMEOUT
> /dev/ttyACM0 [00:07:05.132,690] <inf> event_manager: GPS_EVT_INACTIVE
> /dev/ttyACM0 [00:07:05.138,763] <inf> event_manager: DATA_EVT_DATA_READY
> /dev/ttyACM0 [00:07:05.149,017] <inf> event_manager: DATA_EVT_DATA_SEND
> /dev/ttyACM0 [00:07:05.155,456] <wrn> data_module: No batch data to encode, ringbuffers empty
> /dev/ttyACM0 [00:07:05.171,875] <inf> event_manager: CLOUD_EVT_DATA_ACK
> /dev/ttyACM0 [00:08:00.247,039] <inf> event_manager: APP_EVT_DATA_GET_ALL
> /dev/ttyACM0 [00:08:00.253,509] <inf> event_manager: APP_EVT_DATA_GET - Requested data types (MOD_DYN, BAT, ENV, GNSS)
> /dev/ttyACM0 [00:08:00.270,507] <inf> event_manager: GPS_EVT_ACTIVE
> /dev/ttyACM0 [00:08:00.276,702] <inf> event_manager: SENSOR_EVT_ENVIRONMENTAL_NOT_SUPPORTED
> /dev/ttyACM0 [00:08:00.318,176] <inf> event_manager: MODEM_EVT_MODEM_DYNAMIC_DATA_READY
> /dev/ttyACM0 [00:08:00.359,283] <inf> event_manager: MODEM_EVT_BATTERY_DATA_READY
> ```

```CONFIG_LOG_MAX_LEVEL=4``` demands that the log level for the different modules/libs are set to debug. We have a debug overlay file can be used to enable full logging in the application. ```overlay-debug.conf```

simensrostad commented:
> I get an error when scheduling a FOTA:
> 
> ```
> /dev/ttyACM0 [00:01:07.152,679] <inf> event_manager: CLOUD_EVT_DATA_ACK
> /dev/ttyACM0 [00:01:13.883,728] <err> aws_fota: Error (-7) when trying to start firmware download
> ```
> 
> Did anything of notice change?

I wasn't able to reproduce this. FOTA seems to works fine at my end.

jtguggedal commented:
> Thanks for the update! I might be missing something (pinging @nordic-krch for confirmation), but I believe there are various additional log_strdup() calls that are not necessary and could be removed.

@mbolivar-nordic  Thanks again for taking the time to review. I have done a proper cleanup of the `log_strdup()` usage, and turned out most of them were redundant and are now removed.

coderbyheart commented on applications/asset_tracker_v2/overlay-low-power.conf:
```diff
@@ -0,0 +1,17 @@
+#
```
This file needs a docstring to explain why I would want this, because the sample is already ultra-low-power.

> The Asset Tracker v2 is a real-time configurable ultra-low-power application firmware for the nRF9160 System in Package (SiP).

So, is this `low-power` mode actually consuming more power?

jtguggedal replied:
- Comment added in `overlay-low-power.conf
- README changed to `The Asset Tracker v2 is a real-time configurable ultra-low-power capable application firmware for the nRF9160 System in Package (SiP).`
- Explanation added to the low power overlay description:
```
* :file:`overlay-low-power.conf` - Configuration file to achieve the lowest power consumption, which disables features like LED control and logging that consume extra power.
```

coderbyheart commented on CODEOWNERS:
```diff
@@ -18,6 +18,7 @@
 
 # Applications
 /applications/asset_tracker/              @jhn-nordic @jtguggedal @rlubos
+/applications/asset_tracker_v2/           @simensrostad @jtguggedal @jhn-nordic
```
```suggestion
/applications/asset_tracker_v2/           @simensrostad @jtguggedal @jhn-nordic @coderbyheart
```

jtguggedal replied:
Added

coderbyheart commented on applications/asset_tracker_v2/prj_nrf9160dk_nrf9160ns.conf:
```diff
@@ -0,0 +1,106 @@
+CONFIG_DOWNLOAD_CLIENT=y
+CONFIG_DOWNLOAD_CLIENT_HTTP_FRAG_SIZE_1024=y
+CONFIG_DOWNLOAD_CLIENT_STACK_SIZE=4096
+CONFIG_DOWNLOAD_CLIENT_BUF_SIZE=2300
```
```suggestion
CONFIG_DOWNLOAD_CLIENT_BUF_SIZE=2300
CONFIG_DOWNLOAD_CLIENT_MAX_HOSTNAME_SIZE=128
CONFIG_DOWNLOAD_CLIENT_MAX_FILENAME_SIZE=64
```

coderbyheart replied:
The current defaults do not work with hostnames and filenames generated by the Asset Tracker Cloud Example, therefore we need to provide safe defaults here.

jtguggedal replied:
Added the hostname size config, but keeping the default filename size as it is larger by default (192) and it seems users tend to use large filenames (#3774)

coderbyheart commented on applications/asset_tracker_v2/prj_thingy91_nrf9160ns.conf:
```diff
@@ -0,0 +1,121 @@
+CONFIG_DOWNLOAD_CLIENT=y
+CONFIG_DOWNLOAD_CLIENT_HTTP_FRAG_SIZE_1024=y
+CONFIG_DOWNLOAD_CLIENT_STACK_SIZE=4096
+CONFIG_DOWNLOAD_CLIENT_BUF_SIZE=2300
```
```suggestion
CONFIG_DOWNLOAD_CLIENT_BUF_SIZE=2300
CONFIG_DOWNLOAD_CLIENT_MAX_HOSTNAME_SIZE=128
CONFIG_DOWNLOAD_CLIENT_MAX_FILENAME_SIZE=64
```

nordic-krch commented:
fyi, regarding `log_strdup`. There is a new version of logging on the way that will not require `log_strdup`. At compile time it will detect if it is required and will be applied internally. Draft PR is there (#31535), hopefully it will get into zephyr 2.6.

mbolivar-nordic commented:
@jtguggedal I understand the schedule pressure on this one might mean you do not have time to switch to DEVICE_DT_GET, so please consider that just an FYI for where Zephyr is going that could be handled in a later version of this application.

jtguggedal commented:
> @jtguggedal I understand the schedule pressure on this one might mean you do not have time to switch to DEVICE_DT_GET, so please consider that just an FYI for where Zephyr is going that could be handled in a later version of this application.

@mbolivar-nordic Yes, the schedule is tight, so I'll leave it out of this PR. But we want to show best practices as much as possible, so I'll submit a follow-up PR with this change. Hopefully it can make the release.

jtguggedal commented:
> fyi, regarding `log_strdup`. There is a new version of logging on the way that will not require `log_strdup`. At compile time it will detect if it is required and will be applied internally. Draft PR is there (#31535), hopefully it will get into zephyr 2.6.

@nordic-krch That's a nice improvement!

tejlmand commented on applications/asset_tracker_v2/prj_nrf9160dk_nrf9160ns.conf:
```diff
@@ -0,0 +1,107 @@
+#
```
please merge the common content of `prj_nrf9160dk_nrf9160ns.conf` and `prj_thingy91_nrf9160ns.conf` into a `prj.conf` and then create board fragments for the differences needed under:
`boards/nrf9160dk_nrf9160ns.conf` and `boards/nrf9160dk_nrf9160ns.conf` as described here:
https://docs.zephyrproject.org/latest/guides/kconfig/setting.html#the-initial-configuration

simensrostad reviewed (approved):
LGMT!

tejlmand commented on applications/asset_tracker_v2/src/cloud/cloud_codec/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+zephyr_include_directories(.)
```
why does Zephyr need include directories from source in order to build ?

tejlmand reviewed (commented):
small questions and suggestions, but nothing blocking

jtguggedal commented:
> small questions and suggestions, but nothing blocking

@tejlmand Thanks for reviewing and for the tips. I think we'll leave this PR as is, but will pull in your suggestions in a follow-up PR along with Marti's suggestions on device tree stuff.

tejlmand commented:
> @tejlmand Thanks for reviewing and for the tips. I think we'll leave this PR as is, but will pull in your suggestions in a follow-up PR along with Marti's suggestions on device tree stuff.

Sure, that was why they were just comments, to let you guys decide.
